### ISSUE TITLE
[Truffle] Simple implementation of array

### ISSUE THREAD

**Issue by shishkin** [oracle/graal#1368]
I'm trying to write a simple(r) language with Truffle. I use SL as a reference, but it doesn't have a simple host wrapper container type, e.g. `SLArray` that would wrap JVM `ArrayList<T>` or even `int[]`. My attempts result in Graal complaining that it can't convert my type to Java type. Can somebody please provide a simple example or point me to the right direction of digging?

**Comment by shishkin** [comment]
This code seems to work:

```java
@ExportLibrary(InteropLibrary.class)
public class SLArray implements TruffleObject {

    private final int[] value;

    public SLArray(int[] value) {
        this.value = value;
    }

    @ExportMessage
    static boolean hasArrayElements(SLArray receiver) {
        return true;
    }

    @ExportMessage
    final Object readArrayElement(long index) {
        return value[(int) index];
    }

    @ExportMessage
    final long getArraySize() {
        return value.length;
    }

    @ExportMessage
    final boolean isArrayElementReadable(long index) {
        return 0 <= index && index < getArraySize();
    }
}
```

Maybe worth extending SimpleLanguage with array example?

**Comment by Wooyme** [comment]
There are many things in `SimpleLanguage` that confusing me.
I found it's ok if I just made an ExpressionNode that returns `LinkedList` and of course I can simply add some methods in `SLReadPropertyNode` and `SLWritePropertyNode` to get or add some in the list.
And then I found I can replace all `SLObjectType` to `HashMap`.
So Could anyone tell me what is the use of `TruffleObject`?
I'm really confused about it.

**Comment by boris-spas** [comment]
@shishkin Yeah, SimpleLanguage is a continuous battlefield around whether to add more language features in order to illustrate more, or keep the language simpler to make it more accessible. Over time, it has grown up to really not be a "simple language" anymore. With this in mind, we try to avoid adding things to SL, and encourage people to look at other language implementations once they outgrow SL. Glad you found what you needed!

@Wooyme Not sure what you are asking. Sure, there are ways to avoid using TruffleObject if your use case does not require it, but the TruffleObject javadoc says "Interface for any entity of a Truffle guest language implementations that can be shared across other language implementations.", so if you want your language to interoperate with other truffle languages, you will need TruffleObject.

**Comment by chumer** [comment]
@Wooyme the `TruffleObject` object interface had many purposes in the past. These days, after some APIs got removed, its last purpose is to mark values that are expected to flow through an interpreter.  Between Truffle interpreters we allow primitives like, String, Boolean, Integer and any object that implements TruffleObject. We call such values interop values. Asserting that we see only interop values at the boundary allows to to detect accidental values exposed to other languages.
